tesseract/tesseract_common/include/tesseract_common/clone_cache.h

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file object_cache.h
+ * @brief
+ *
+ * @author Levi Armstrong
+ * @author Matthew Powelson
+ * @date November 15, 2020
+ * @version TODO
+ * @bug No known bugs
+ *
+ * @copyright Copyright (c) 2020, Southwest Research Institute
+ *
+ * @par License
+ * Software License Agreement (Apache License)
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TESSERACT_COMMON_OBJECT_CACHE_H
+#define TESSERACT_COMMON_OBJECT_CACHE_H
+
+#include <tesseract_common/macros.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_PUSH
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <console_bridge/console.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_POP
+
+namespace tesseract_common
+{
+template <typename T>
+class CloneCache
+{
+public:
+  CloneCache(T original, const long& cache_size = 5) : original_(original), cache_size_(cache_size)
+  {
+    for (long i = 0; i < cache_size; i++)
+      createClone();
+  }
+
+  const std::shared_ptr<T>& operator->() { return original_; }
+
+  std::shared_ptr<T> clone()
+  {
+    std::unique_lock<std::shared_mutex> lock(cache_mutex_);
+
+    std::shared_ptr<T> t;
+    if (cache_revision_ != original_->getRevision() || cache_.empty())
+    {
+      // TODO: Check if original_->update exists
+      if (false)
+      {
+        original_->update(cache_.back());
+        t = cache_.back();
+        cache_.pop_back();
+      }
+      else
+        t = original_->clone();
+    }
+    else
+    {
+      t = cache_.back();
+      cache_.pop_back();
+    }
+
+    return t;
+  }
+
+  /**
+   * @brief Set the cache size used to hold tesseract objects for motion planning
+   * @param size The size of the cache.
+   */
+  void setCacheSize(long size)
+  {
+    std::unique_lock<std::shared_mutex> lock(cache_mutex_);
+    cache_size_ = static_cast<std::size_t>(size);
+  }
+
+  /**
+   * @brief Get the cache size used to hold tesseract objects for motion planning
+   * @return The size of the cache.
+   */
+  long getCacheSize() const { return static_cast<long>(cache_size_); }
+
+  /** @brief If the environment has changed it will rebuild the cache of tesseract objects */
+  void refreshCache()
+  {
+    // TODO: Check if original_->update exists and use that if it does
+
+    std::unique_lock<std::shared_mutex> lock(cache_mutex_);
+    std::shared_ptr<T> thor;
+    {
+      auto lock = original_->getMutex();
```

Comment:
I could see this being need in case where the original and cache are being used in a multi threaded application but  it maybe better for the developer to handle this outside of this class keeping the cache class simpler where it only worries about its internal data being access in a multi threaded case. What do you think?

Reply:
Yeah that makes sense. The goal was to make a CloneCache<T> object be somewhat interchangeable with a T object. So the user would be responsible for locking it when changing original_